File: cpp/0037_SudokuSolver.cpp


class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        solver(board,0,0);
        
    }
    bool solver(vector<vector<char>>& board,int i,int j)
    {
        if(i==9) return true;
        if(j==9) return solver(board,i+1,0);
        if(board[i][j]!='.') return solver(board,i,j+1);
        for(char c='1';c<='9';c++)
        {
            if(isValid(board,i,j,c))
            {
                board[i][j]=c;

            }
            else
                continue;
            if (solver(board, i, j + 1)) 
                return true;
            board[i][j] = '.';
        }
        return false;


    }
    bool isValid(vector<vector<char>>& board,int i,int j,char c)
    {
        int row = i - i % 3, col = j - j % 3;
        for(int k=0;k<9;k++)
        {
            if(board[i][k]==c )
                return false;
            if(board[k][j]==c )
                return false;
            if(board[k/3+row][k%3+col]==c)
                return false;
        }
        
        return true;

    }
};

File: python/0037_SudokuSolver.py
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        self.solver(board,0,0)

    def solver(self,board,i,j):
        if i==9:
            return True;
        if j==9:
            return self.solver(board,i+1,0)
        if board[i][j]!='.':
            return self.solver(board,i,j+1)

        for k in range(1,10):
            if not self.isValid(board,i,j,str(k)):
                continue
            board[i][j]=str(k)
            if(self.solver(board,i,j+1)):
                return True
            board[i][j]='.'
        return False        

    def isValid(self,board,i,j,c):
        row,col=i-i%3,j-j%3
        for k in range(9):
            if board[i][k]==c:
                return False
            if board[k][j]==c:
                return False
            if board[row+k/3][col+k%3]==c:
                return False

        return True

